cmd/moby/docker.go

```diff
@@ -20,91 +20,89 @@ import (
 	"golang.org/x/net/context"
 )
 
-func dockerRun(args ...string) ([]byte, error) {
-	log.Debugf("docker run: %s", strings.Join(args, " "))
-	docker, err := exec.LookPath("docker")
-	if err != nil {
-		return []byte{}, errors.New("Docker does not seem to be installed")
-	}
-	args = append([]string{"run", "--rm", "--log-driver=none"}, args...)
-	cmd := exec.Command(docker, args...)
-
-	stderrPipe, err := cmd.StderrPipe()
-	if err != nil {
-		return []byte{}, err
-	}
-
-	stdoutPipe, err := cmd.StdoutPipe()
-	if err != nil {
-		return []byte{}, err
-	}
+func dockerRunInput(input io.Reader, image string, args ...string) ([]byte, error) {
+	log.Debugf("docker run (input): %s %s", image, strings.Join(args, " "))
 
-	err = cmd.Start()
+	cli, err := dockerClient()
 	if err != nil {
-		return []byte{}, err
+		return []byte{}, errors.New("could not initialize Docker API client")
 	}
 
-	stdout, err := ioutil.ReadAll(stdoutPipe)
-	if err != nil {
-		return []byte{}, err
-	}
+	origStdin := os.Stdin
+	defer func() {
+		os.Stdin = origStdin
+	}()
 
-	stderr, err := ioutil.ReadAll(stderrPipe)
+	// write input to a file
+	rawInput, err := ioutil.ReadAll(input)
 	if err != nil {
-		return []byte{}, err
+		return []byte{}, errors.New("could not read input")
 	}
-
-	err = cmd.Wait()
-	if err != nil {
-		return []byte{}, fmt.Errorf("%v: %s", err, stderr)
+	tmpInputFileName := "tmpInput"
+	if err := ioutil.WriteFile(tmpInputFileName, rawInput, 0644); err != nil {
+		return []byte{}, errors.New("could not stage input to file")
 	}
-
-	log.Debugf("docker run: %s...Done", strings.Join(args, " "))
-	return stdout, nil
-}
-
-func dockerRunInput(input io.Reader, args ...string) ([]byte, error) {
-	log.Debugf("docker run (input): %s", strings.Join(args, " "))
-	docker, err := exec.LookPath("docker")
+	inputFile, err := os.Open(tmpInputFileName)
 	if err != nil {
-		return []byte{}, errors.New("Docker does not seem to be installed")
+		return []byte{}, errors.New("could not open input file for container stdin")
 	}
-	args = append([]string{"run", "--rm", "-i"}, args...)
-	cmd := exec.Command(docker, args...)
-	cmd.Stdin = input
+	defer os.Remove(tmpInputFileName)
 
-	stderrPipe, err := cmd.StderrPipe()
-	if err != nil {
-		return []byte{}, err
-	}
+	os.Stdin = inputFile
```

Comment:
`io.Copy(resp.Conn, inputFile)` I think is what you need to to the stdin of the container.